File: problems/CSES/Graph_Algorithms/mail_delivery/mail_delivery.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, m = map(int, input().split())

    graph = [[] for _ in range(n)]
    degree = [0] * n

    for edge_id in range(m):
        a, b = map(int, input().split())
        a -= 1
        b -= 1

        graph[a].append((b, edge_id))
        graph[b].append((a, edge_id))

        degree[a] += 1
        degree[b] += 1

    for v in range(n):
        if degree[v] % 2:
            print("IMPOSSIBLE")
            return

    used = [False] * m
    index = [0] * n

    stack = [0]
    path = []

    while stack:
        v = stack[-1]

        while index[v] < len(graph[v]) and used[graph[v][index[v]][1]]:
            index[v] += 1

        if index[v] == len(graph[v]):
            path.append(v)
            stack.pop()

        else:
            u, edge_id = graph[v][index[v]]
            index[v] += 1

            if not used[edge_id]:
                used[edge_id] = True
                stack.append(u)

    if len(path) != m + 1:
        print("IMPOSSIBLE")
        return

    path.reverse()

    print(*[v + 1 for v in path])

main()

File: problems/CSES/Graph_Algorithms/mail_delivery/mail_delivery.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<vector<pair<int, int>>> graph(n);
    vector<int> degree(n, 0);

    for (int edge_id = 0; edge_id < m; ++edge_id) {
        int a, b;
        cin >> a >> b;

        a--;
        b--;

        graph[a].pb({b, edge_id});
        graph[b].pb({a, edge_id});

        degree[a]++;
        degree[b]++;
    }

    for (int v = 0; v < n; ++v) {
        if (degree[v] % 2) {
            cout << "IMPOSSIBLE\n";
            return 0;
        }
    }

    vector<bool> used(m, false);
    vector<int> index(n, 0);

    vector<int> stack = {0};
    vector<int> path;

    while (!stack.empty()) {
        int v = stack.back();

        while (
            index[v] < (int)graph[v].size() &&
            used[graph[v][index[v]].second]
        ) {
            index[v]++;
        }

        if (index[v] == (int)graph[v].size()) {
            path.pb(v);
            stack.pop_back();
        } else {
            auto [u, edge_id] = graph[v][index[v]];

            index[v]++;

            if (!used[edge_id]) {
                used[edge_id] = true;
                stack.pb(u);
            }
        }
    }

    if ((int)path.size() != m + 1) {
        cout << "IMPOSSIBLE\n";
        return 0;
    }

    reverse(path.begin(), path.end());

    for (int v : path) {
        cout << v + 1 << ' ';
    }
    cout << '\n';

    return 0;
}
